### PR TITLE
Add groupby to dask.dataframe.Series

### LINKED ISSUE #575: Series.groupby
We currently lack a `Series.groupby` method.  Fixing this probably involves a slight refactoring of the current groupby code.

```python
In [1]: import pandas as pd
In [2]: s = pd.Series([1, 2, 2, 1, 1])
In [3]: s.groupby(s).count()
Out[3]: 
1    3
2    2
dtype: int64

In [4]: import dask.dataframe as dd
In [5]: ss = dd.from_pandas(s, npartitions=2)
In [6]: ss.groupby(ss).count()
AttributeError: 'Series' object has no attribute 'groupby'
```

### PR DESCRIPTION
**Pull request by cowlicks**
closes #575

### CODE REVIEW

**Comment by mrocklin** [comment]
+1

**Comment by mrocklin** [comment]
Might be nice to wait on merging to see if @sinhrks has any thoughts on this.

**Comment by cowlicks** [comment]
rebased

**Comment by sinhrks** [comment]
Looks nice!

- Better to validate ``index`` is a Series which has the same division, like [GroupBy](https://github.com/blaze/dask/blob/master/dask/dataframe/core.py#L1561) (or ``use _maybe_partition``)
- Can you add a test with different ``_name``, like ``ss.groupby(ss + 1)``

File: dask/dataframe/core.py [hunk 1]
```diff
@@ -1619,12 +1626,19 @@ class SeriesGroupBy(_GroupBy):
...
         self.index = index
         self.key = key
         self.kwargs = kwargs
 
+        if not isinstance(index, Series):
+            raise TypeError("A dask Series must be used as the index for a "
+                            "Series groupby.")
+        if not df.divisions == index.divisions:
+            raise ValueError("The Series and index of the groupby must have "
+                             "the same divisions.")
```
Comment (mrocklin) [inline dask/dataframe/core.py]: Should probably be a `NotImplementedError`

File: dask/dataframe/tests/test_dataframe.py [hunk 2]
```diff
@@ -1881,3 +1881,28 @@ def test_series_groupby_propagates_names():
...
+    assert eq(dask_group.min(), pd_group.min())
+    assert eq(dask_group.max(), pd_group.max())
+
+    pd_group2 = s.groupby(s + 1)
+    dask_group2 = ss.groupby(ss + 1)
+
+    assert eq(dask_group2.count(), pd_group2.count())
+    assert eq(dask_group2.sum(), pd_group2.sum())
+    assert eq(dask_group2.min(), pd_group2.min())
+    assert eq(dask_group2.max(), pd_group2.max())
```
Comment (mrocklin) [inline dask/dataframe/tests/test_dataframe.py]: Can these two groups of asserts be combined with a for loop?

**Pull request merged**